soc: nuvoton: numaker: add support for m2l31x series

cyliangtw opened the pull request:
This PR is to add Nuvoton NuMaker M2L31X SoC series.

This PR depends on https://github.com/zephyrproject-rtos/hal_nuvoton/pull/9

fabiobaltieri commented on drivers/flash/soc_flash_numaker_rmc.c:
```diff
@@ -0,0 +1,275 @@
+	struct flash_numaker_data *dev_data = dev->data;
+	uint32_t rc = 0;
+	unsigned int key;
+	int page_nums = (len / RMC_FLASH_PAGE_SIZE);
```
same

fabiobaltieri commented on drivers/flash/soc_flash_numaker_rmc.c:
```diff
@@ -0,0 +1,275 @@
+	uint32_t addr = dev_data->flash_block_base + offset;
+
+	/* return SUCCESS for len == 0 (required by tests/drivers/flash) */
+	if (!len) {
```
nit: should probably be `if (len == 0) {` since len is not a boolean, same below

fabiobaltieri commented on drivers/flash/soc_flash_numaker_rmc.c:
```diff
@@ -0,0 +1,275 @@
+	uint32_t addr = dev_data->flash_block_base + offset;
+	int block_size = flash_numaker_parameters.write_block_size;
+	int blocks = (len / flash_numaker_parameters.write_block_size);
+	uint8_t *pu8_data = (uint8_t *)data;
```
do you need the casting? you are discarding the `const`

cyliangtw replied:
Thanks of your advice, const is mandatory.

cyliangtw commented on dts/arm/nuvoton/m2l31x.dtsi:
```diff
@@ -0,0 +1,228 @@
+			compatible = "nuvoton,numaker-scc";
+			reg = <0x40000200 0x100>;
+			#clock-cells = <0>;
+			/* hxt = "enable"; */
+			/* lxt = "enable"; */
```
already support and verified on board.
it depends on board built-in crystal oscillator or not, so drop & move to board.dts .

cyliangtw commented on dts/arm/nuvoton/m2l31x.dtsi:
```diff
@@ -0,0 +1,228 @@
+			/* lxt = "enable"; */
+			clk-pclkdiv = <(NUMAKER_CLK_PCLKDIV_APB0DIV_DIV2 |
+					 NUMAKER_CLK_PCLKDIV_APB1DIV_DIV2)>;
+			core-clock = <72000000>;
```
OK

cyliangtw commented on soc/nuvoton/numaker/soc.yml:
```diff
@@ -4,3 +4,6 @@ family:
     - name: m467
+  - name: m2l31x
+    socs:
+    - name: m2l31ki
```
after review this M2L31 series, the different parts would cover different SRAM/ROM size, even SPI/I2C bodies.
So, you suggestion is much helpful and implemented specific dtsi for different category of parts at commit  3ab1e00a09e6.

fabiobaltieri commented on dts/arm/nuvoton/m2l31kid.dtsi:
```diff
@@ -0,0 +1,22 @@
+
+#include <mem.h>
+#include <nuvoton/m2l31x.dtsi>
+/ {
```
nit: you'd normally have a blank line after the includes

nordicjm reviewed (dismissed):
Build system/hwmv2 files are OK

fabiobaltieri commented on drivers/flash/soc_flash_numaker_rmc.c:
```diff
@@ -0,0 +1,275 @@
+	uint32_t addr = dev_data->flash_block_base + offset;
+
+	/* return SUCCESS for len == 0 (required by tests/drivers/flash) */
+	if (!len) {
```
nit: `len == 0`

fabiobaltieri commented on drivers/flash/soc_flash_numaker_rmc.c:
```diff
@@ -0,0 +1,275 @@
+	const uint8_t *pu8_data = (const uint8_t *)data;
+
+	/* return SUCCESS for len == 0 (required by tests/drivers/flash) */
+	if (!len) {
```
`len == 0`

cyliangtw replied:
@fabiobaltieri ,
Thanks and finished the update.

fabiobaltieri reviewed (commented):
cool, just a couple more nitpicks from me

cyliangtw committed:
soc: nuvoton: numaker: add support for m2l31x series

Add initial support for nuvoton numaker m2l31x SoC series
including basic init.

Signed-off-by: cyliang tw <cyliang@nuvoton.com>

cyliangtw commented:
Hi @nordicjm , @fabiobaltieri,

Rebase to sync with main branch  to resolve merge conflict of CMakeLists.txt & Kconfig of drivers/flash.

cyliangtw committed:
drivers: flash: support for Nuvoton numaker series RMC

Add Nuvoton numaker series flash memory controller(RMC) with erase,
 read & write features of soc-flash.

Signed-off-by: cyliang tw <cyliang@nuvoton.com>

nordicjm reviewed (approved):
Build system files OK
